packages/experiment-tag/src/experiment.ts

```diff
@@ -8,443 +8,675 @@ import {
+          // @ts-ignore
+          user.currentUrl = currentUrl;
+        }
+        this.experimentClient.setUser(user);
+      } else {
+        this.experimentClient.setUser({
+          ...existingUser,
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore
+          currentUrl: currentUrl,
+        });
+      }
+    }
   }
-  // apply remote variants - if fetch is unsuccessful, fallback order: 1. localStorage flags, 2. initial flags
-  applyVariants(globalScope.webExperiment.all(), remoteFlagKeys);
-};
-
-const applyVariants = (
-  variants: Variants,
-  flagKeys: Set<string> | undefined = undefined,
-) => {
-  if (Object.keys(variants).length === 0) {
-    return;
+
+  /**
+   * Set the previous URL for tracking back/forward navigation. Set previous URL to prevent infinite redirection loop
+   * in single-page applications.
+   * @param url
+   */
+  public setPreviousUrl(url: string) {
+    this.previousUrl = url;
   }
-  const globalScope = getGlobalScope();
-  if (!globalScope) {
-    return;
+
+  /**
+   * Apply evaluated variants to the page.
+   * @param flagKeys
+   */
+  public applyVariants(flagKeys: string[] | undefined = undefined) {
+    const variants = this.experimentClient?.all() || {};
+    if (Object.keys(variants).length === 0) {
+      return;
+    }
+    const globalScope = getGlobalScope();
+    if (!globalScope) {
+      return;
+    }
+    const currentUrl = urlWithoutParamsAndAnchor(globalScope.location.href);
+    // Initialize the cache if on a new URL
+    if (!this.urlExposureCache?.[currentUrl]) {
+      this.urlExposureCache = {};
+      this.urlExposureCache[currentUrl] = {};
+    }
+    for (const key in variants) {
+      if (flagKeys && !flagKeys.includes(key)) {
+        continue;
+      }
+      const variant = variants[key];
+      const isWebExperimentation = variant.metadata?.deliveryMethod === 'web';
+      if (isWebExperimentation) {
+        const shouldTrackExposure =
+          (variant.metadata?.['trackExposure'] as boolean) ?? true;
+        // if payload is falsy or empty array, consider it as control variant
+        const payloadIsArray = Array.isArray(variant.payload);
+        const isControlPayload =
+          !variant.payload || (payloadIsArray && variant.payload.length === 0);
+        if (shouldTrackExposure && isControlPayload) {
+          this.exposureWithDedupe(key, variant);
+          continue;
+        }
+
+        if (payloadIsArray) {
+          this.handleVariantAction(key, variant);
+        }
+      }
+    }
   }
-  const currentUrl = urlWithoutParamsAndAnchor(globalScope.location.href);
-  // Initialize the cache if on a new URL
-  if (!urlExposureCache?.[currentUrl]) {
-    urlExposureCache = {};
-    urlExposureCache[currentUrl] = {};
+
+  /**
+   * Revert mutations applied by the experiment.
+   * @param flagKeys
+   */
+  public revertMutations(flagKeys: string[] | undefined = undefined) {
```

Comment:
Change to object input, `RevertVariantsOptions`, add `flagKeys` as optional field.